### DISCUSSION TITLE
TSC Codec Adapter - Minutes and Outcomes

### DISCUSSION THREAD

**Discussion by lgirdwood** [thesofproject/sof#7391] [Team Posts]
**Attendees**
@dgreid @dbaluta @lgirdwood @mwasko @beatabaranowska @bzhg @plbossart @harshapriya-n @cujomalainey Guests @mrajwa @stolx

Added initial priority to items that I think need done first. Please discuss here though and this can be updated.

**Minutes**
1) Registration - Array based codec registration today with secondary ID , codec adatper contains helper logic and it's own UUID.
- [ ] Remove secondary IDs and use GUIDs - **P1**
- [ ] Use linker for registration like other componnets (voume, mux, src etc) **P1**
- [ ] Codec adapter helper APIs to be base firmware APIs.

2) Multiplexer - A single container component that can contain N codecs and switch between them at runtime. Uses a single pipeline.
- [ ] Use GUID to select codec
- [ ] Select codec at hw_params() time.

3) Codec Helper APIs - Helpers not in base or adapter today that should be added for convenience and reducing runtime logic and effort.
- [ ] Config structure to define sets of runtime parameters so that runtime code is not needed in most cases.
- [ ] Stack size limits need to be known.

4) Codec configuration - two types of configuration data exist today, pipeline topology and runtime configuration data. Data has been seen to be out of order depending on order in topology - TBC.
- [ ] Always run apply_config callback regardless if codec is prepared and regardless of config type
- [ ] runtime config set before hw_params time (anytime after adapter inited)

5) Codec state machine - another set of states exist with adapter.
 - [ ] Codec adapter to use standard component state machine (can add states if needed). **P1**
 
6) Compressed streams - some codecs are initialised by reading stream data rather than at hw_params time. Idea to be discussed, but 2 connected pipelines would work.
 - [ ] Upstream component pipeline should have a VBR or not PCM flag with it's hw_params (i.e. tells upstream code to just copy if they support non PCM)
 - [ ] Component needs logic and to get hw_params fram start of data and should initiate the downstream pipeline with the params once they are known.

**Comment by cujomalainey** [comment]
1 is being tackled in #3881 it should make 2 easier to accomplish

**Comment by cujomalainey** [comment]
I think minutes may have captures #4 changes incorrectly. It captures the current state of the config system. Not the requested changes. Changes requested are for configs to push to adapter after init, current system is after hw params to align with configs pushed

**Comment by lgirdwood** [comment]
@cujomalainey appologies, I took the current state in the minutes. Can you update to the desired state. I am working on parts of this now...

**Comment by cujomalainey** [comment]
Updated, @all please check 4 for errors or if my memory does not recall the resolution correctly (im still a bit groggy from time change 😛 )

**Comment by lgirdwood** [comment]
@ranj063 fyi
